src/driver/amdxdna/aie2_pci.c

```
@@ -1327,44 +1308,95 @@ static int aie2_query_ctx_status_array(struct amdxdna_client *client,
 		srcu_read_unlock(&tmp_client->ctx_srcu, idx);
 	}
 
-	min = min(args->element_size, sizeof(*tmp));
-	for (i = 0; i < hw_i; i++) {
-		if (copy_to_user(&buf[i], &tmp[i], min)) {
-			ret = -EFAULT;
-			break;
-		}
+	if (pid && ctx_id && !hw_i) {
+		XDNA_ERR(xdna, "Invalid context ID %d for PID %lld", ctx_id, pid);
+		ret = -EINVAL;
 	}
 
 	aie2_mgmt_buff_free(&mgmt_hdl);
-	kfree(tmp);
-	args->element_size = min;
-	args->num_element = hw_i;
 	return ret;
 }
 
-static int aie2_get_info_array(struct amdxdna_client *client,
-			       struct amdxdna_drm_get_info_array *args)
+static int aie2_get_array(struct amdxdna_client *client, struct amdxdna_drm_get_array *args)
 {
+	struct amdxdna_drm_hwctx_entry __user *buf;
 	struct amdxdna_dev *xdna = client->xdna;
-	int ret;
+	struct amdxdna_drm_hwctx_entry *tmp;
+	int ctx_limit, ctx_cnt, min, i, ret;
+	u32 buf_size;
 
 	ret = amdxdna_pm_resume_get(xdna);
 	if (ret)
 		return ret;
 
 	mutex_lock(&xdna->dev_lock);
-	mutex_lock(&xdna->dev_handle->aie2_lock);
+
+	ctx_limit = aie2_rq_context_limit(&xdna->dev_handle->ctx_rq);
+	WARN_ON(ctx_limit > AMDXDNA_MAX_NUM_ELEMENT);
+	ctx_cnt = aie2_rq_active_context(&xdna->dev_handle->ctx_rq);
```

Move to aie2_query_ctx_status_array?

---

Moved the above lines to the `DRM_AMDXDNA_HW_CONTEXT_ARRAY` case. As args->num_element needs to be greater than or equal to ctx_cnt. If not, should fail early.